src/parseMessage.h

```diff
@@ -0,0 +1,44 @@
+#ifndef PARSE_MESSAGE_H
+#define PARSE_MESSAGE_H
+
+#include <stdint.h>
+#include <stddef.h>
+
+typedef struct MessageParser {
+    uint8_t* buffer;
+    size_t buffer_length;
+} MessageParser;
+
+typedef struct Pubkey {
+    uint8_t data[32];
+} Pubkey;
+
+typedef struct Blockhash {
+    uint8_t data[32];
+} Blockhash;
+
+typedef struct Instruction {
+    uint8_t program_id_index;
+    uint8_t* accounts;
+    size_t accounts_length;
+    uint8_t* data;
+    size_t data_length;
+} Instruction;
+
+typedef struct MessageHeader {
+    uint8_t num_required_signatures;
+    uint8_t num_readonly_signed_accounts;
+    uint8_t num_readonly_unsigned_accounts;
+} MessageHeader;
+
+int parse_length(MessageParser* parser, size_t* value);
+
+int parse_message_header(MessageParser* parser, MessageHeader* header);
+
+int parse_pubkeys(MessageParser* parser, Pubkey** pubkeys, size_t* pubkeys_length);
+
+int parse_blockhash(MessageParser* parser, Blockhash** blockhash);
+
+int parse_instruction(MessageParser* parser, Instruction* instruction);
+
```

Comment:
Could combine the above into a `parse_message()`.  I guess that could result in an unwieldy structure...

Reply:
Right, that's exactly the issue.  By stopping at `parse_instruction()`, it becomes the job of the caller to iterate over the instructions one by one and extract the information they need.  I'm trying to keep this as a zero-copy, zero-alloc parser.  It's not obvious to me what the function signature of `parse_message` would be.  What's the thinking on VLAs these days?

Reply:
You can probably do something similar to what we do currently with `SolKeyedAccount` for C BPF programs.  `Message` would be more complicated but I think the same concepts apply.  A collection of pointers, lengths, and non-pointer members.

```c
/**
 * Keyed Account
 */
typedef struct {
  SolPubkey *key;        /** Public key of the account */
  bool is_signer;        /** Transaction was signed by this account's key? */
  bool is_writable;      /** Is the account writable? */
  uint64_t *lamports;    /** Number of lamports owned by this account */
  uint64_t userdata_len; /** Length of data in bytes */
  uint8_t *userdata;     /** On-chain data within this account */
  SolPubkey *owner;      /** Program that owns this account */
} SolKeyedAccount;
```

Where for example `SolPubkey` is:

```c
/**
 * Size of Public key in bytes
 */
#define SIZE_PUBKEY 32

/**
 * Public key
 */
typedef struct {
  uint8_t x[SIZE_PUBKEY];
} SolPubkey;
```

Reply:
I see, thanks.  P.S. "userdata" !?

Reply:
Heh, yeah, needs updating, will do :-)